### DISCUSSION TITLE
GUI version coming soon!

### DISCUSSION THREAD

**Discussion by YarosMallorca** [YarostheLaunchpadder/MissionPlanner-to-Litchi#9] [Announcements]
## I'm working on a more complete yet simple version of this app, coming soon!
New features include: 
- More Mission Planner Commands translated
- Easy-to-use interface, no command line
- Open and Save to selected folders
- MacOS version

**Comment by JoeKae** [comment]
## GUI

The tkinter GUI I submitted was just a quick way to do things. If you want a more polished look, we should look at other frameworks.

## Platform dependency

At some point, you might want to provide a Docker image to get a platform-independent "executable."

## Features

I have a suggestion that I would like to see.

### Terrain Follow

#### What it is

Some expensive tools offer the ability to do "Terrain Follow" which tracks the elevation of the terrain in more detail than the mission planner does.

#### Where we are now

The Mission Planner sets each waypoint relative to the terrain and interpolates between them, but ignores the actual elevation changes between waypoints.
See the example [terrain_follow_test.zip] (https://github.com/YarostheLaunchpadder/MissionPlanner-to-Litchi/files/9941398/terrain_follow_test.zip).

#### What it should be

It would be a nice feature to calculate the elevation changes between waypoints. When a certain threshold is exceeded, a new set of waypoints should be generated to compensate for the elevation differences.
Thresholds could be absolute values in meters or relative values in percent of altitude. Mapping guides recommend that a 20% change in altitude relative to flight altitude should be the maximum.

---

I know this is not in the scope of this application, but it would be nice if it was.
I would be happy to assist whenever i can find the spare time. Maybe you can add me as an colab?

**Reply by YarosMallorca** [reply]
Yes sure!
I have looked at Flutter for the UI, it uses Dart as backend language. But then I thought that translating the code to Dart might be daunting, but it might be worth it! What do you think? Flutter can compile to all platforms like Windows, macOS, Linux, and even mobile! Though the problem is that Mission Planner is Windows-only software. 

Sure, I can add you as colab :)

**Reply by JoeKae** [reply]
I would suggest to stay with python.
The reasons are:
* Python is widely used, so people will be more likely to contribute.
* The codebase, while small, is python. So there would be no need to port it to another language.
* Most frameworks are lightweight.

**You as the maintainer have to make the call though.**
I do not see real use cases for mobile devices because, as you mentioned already, Mission Planner is a Windows software.

I would suggest to build an appealing GUI in [PyQt](https://doc.qt.io/qtforpython/). Python and its dependencies are easy to manage using virtual environments ([requirements.txt](https://docs.python.org/3/tutorial/venv.html), [pipenv](https://pipenv.pypa.io/en/latest/)).
Also we can use githubs [dependabot](https://github.com/dependabot).

Since we would rely on a GUI, we cannot use Docker to bundle the app, because we cannot spawn a GUI from Docker in Windows.
Any linux user will not have a hard time running the app anyway. Windows users, on the other hand, are used to .exe files which we can create utilizing [pyinstaller](https://pyinstaller.org/en/stable/).

---

What do you think about the terrain follow idea?

**Reply by YarosMallorca** [reply]
Sure! It would be nice to implement the terrain follow feature, and I don't think it's that hard either, it's just adapting the code to enable the "above ground" parameter at each waypoint in the CSV. 

Yes, sure let's continue using Python for this project then. No need to migrate the codebase to another platform.
I have never used PyQt, but it seems nice! I have only done Python GUIs in Tkinter and Kivy, but I would love to try PyQt for this project as by your suggestion.

**Reply by YarosMallorca** [reply]
Also, what are your thoughts on using something like [CustomTkinter](https://github.com/TomSchimansky/CustomTkinter) for the UI. Or is PyQt better because of its visual builder tool?

**Reply by JoeKae** [reply]
> it's just adapting the code to enable the "above ground" parameter at each waypoint in the CSV.

It is more than that. Just setting the AGL flag would interpolate the altitude linear between waypoints, ignoring elevation changes between the waypoints.

---

I never really worked with PyQt either, but it would be a nice reason to check it out.

**Reply by YarosMallorca** [reply]
Right, but also in Mission Planner there is an Above Ground flag too, that calculates the elevation using Google Maps data. So maybe that will work.

**Reply by JoeKae** [reply]
Can you please show me a screenshot? I do not find the flag.

**Reply by YarosMallorca** [reply]
Do this before generating the survey. Make sure the height in the text box highlighted in the screenshot is the same as in the survey creator window.

![image](https://user-images.githubusercontent.com/54041533/200196933-6fd202ff-3dfe-4d72-ab48-d4afbbec4d60.png)

**Reply by JoeKae** [reply]
Ok, now I know what you are talking about. I usually select terrain so it does use the terrain elevation model.
The problem is that this is not terrain follow. It is simple AGL.

### Example
Have a look at this example here:
![Example](https://user-images.githubusercontent.com/52297861/200197236-cc96a2db-e360-49fa-bd35-b422f796f8db.PNG)

This is a flight plan that results from using those settings. All the waypoints are adjusted to the terrain elevation, but the terrain elevation between the waypoints is ignored.
Terrain follow would check the elevation on the whole flight path and create more waypoints to adjust the height if needed.

A terrain follow flight plan should look more like the ground projection in the image. Of course the limiting factor is ground elevation data resolution, but you can create a [DTM, DEM or DSM](https://support.plexearth.com/hc/en-us/articles/4642425453201-Elevation-Modeling-the-differences-between-DTM-DSM-DEM) first and use that as a datasource.

**Reply by JoeKae** [reply]
Have a look at this. This does exactly what i explain:
https://github.com/JMG30/flight_planner

**Reply by JoeKae** [reply]
As of PR #14 it is possible to use the `DO_DIGICAM_CONTROL`Trigger Method (camera options tab). That way we get a waypoint for each photo. Because it is set to AGL we get a weak terrain follow functionality. See image:
![agl_for_each_photo](https://user-images.githubusercontent.com/52297861/202852229-b87f5e89-4117-49d0-93b7-927f941643d9.JPG)

While this is no real terrain follow it is a good first step.

**Comment by JoeKae** [comment]
Add file drop to the gui wishlist. So we can just drag&drop the `.waypoints` files into the gui for them to be converted.
This [example for PYQT5 with text](https://zetcode.com/gui/pyqt5/dragdrop/) could be a good starting point.

**Comment by sgraham3** [comment]
Not sure if if this should go here or if its possible without a lot of work. I see there are some python scripts in the Mission Planner directory. (Idea)The ability to install MP-to-Litchi as a plugin in Mission Planer and simply  export to litchi from MP.

**Reply by JoeKae** [reply]
I think you should ask here:
https://github.com/ArduPilot/MissionPlanner

**Reply by sgraham3** [reply]
Ok maybe I should have not added to this thread. I was just floating the idea instead of creating a new gui intergrading the existing script in MP.
